### PR TITLE
STALLED: db: add BadgerDB

### PR DESCRIPTION
**Pull request by odeke-em**
Updates https://github.com/tendermint/tmlibs/issues/30

Adds a backend for BadgerDB as requested at
https://github.com/tendermint/tmlibs/issues/30#issue-241573231
and tests for it along.

BadgerDB has a few problems:
a) It doesn't have a BatchDelete, our fallback is thus Delete
because if do DeleteAsync, by the time batch.Write is invoked,
we can't be sure that the deletions have ran
b) Retrieving values from iterators requires a secondary allocation
equal to the length of the current key.
See https://godoc.org/github.com/dgraph-io/badger#KV.NewIterator
c) It seems to have arbitrary options for the size of the
log file varying between 10MB to 2GB. Currently with our
DB API, we can't pass in the required file size unless we
invoke:
```go
db.NewBadgerDBWithOptions(&db.Options{
  ValueLogFileSize: desiredFileSize,
})
```

Anyways, I've arbitrarily set the default value log file size to 1GB
given that Tendermint does a whole lot of heavy lifting with
keyvalue stores <-- cite my sources

### CODE REVIEW

File: db/badger_db.go [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+//  if err != nil {
+//	log.Fatalf("badgerDB initialization: %v", err)
+//  }
+//  defer db.Close()
+//
+//  For the asynchronous set
+//  db.Set([]byte("foo"), []byte("bar"))
+//  db.SetSync([]byte("true"), []byte("tendermint"))
+//  db.Delete([]byte("true"))
+//  db.Deletesync([]byte("true"))
```
Comment (melekes) [inline db/badger_db.go]: `DeleteSync`

File: db/badger_db.go [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+	_MB = 1024 * _KB
+	_GB = 1024 * _MB
+)
+
+// NewBadgerDB creates a Badger key-value store backed to the
+// directory dir supplied. If dir does not exist, we create it.
+func NewBadgerDB(dbName, dir string) (*BadgerDB, error) {
+	// BadgerDB doesn't expose a way for us to
+	// create  a DB with the user's supplied name.
+	if err := os.MkdirAll(dir, 0755); err != nil {
```
Comment (melekes) [inline db/badger_db.go]: Don't you think `MkdirAll ` belongs to `NewBadgerDBWithOptions` function?

File: db/badger_db.go [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+		return nil, err
+	}
+	opts := new(badger.Options)
```
Comment (melekes) [inline db/badger_db.go]: what if we just do `opts := badger.DefaultOptions`?

File: db/badger_db.go [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+type BadgerDB struct {
+	kv *badger.KV
+}
+
+var _ DB = (*BadgerDB)(nil)
+
+func (b *BadgerDB) Get(key []byte) []byte {
+	valueItem := new(badger.KVItem)
+	if err := b.kv.Get(key, valueItem); err != nil {
+		// Unfortunate that Get can't return errors
```
Comment (melekes) [inline db/badger_db.go]: 💯

File: db/badger_db.go [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+		// TODO: Propose allowing DB's Get to return errors too.
+		common.PanicCrisis(err)
+	}
+	var valueSave []byte
+	err := valueItem.Value(func(origValue []byte) error {
+		// TODO: Decide if we should just assign valueSave to origValue
```
Comment (melekes) [inline db/badger_db.go]: so why can't we just assign?

File: db/badger_db.go [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+	})
+	bb.entriesMu.Unlock()
+}
+
+// Unfortunately Badger doesn't have a batch delete
+// The closest that we can do is do a delete from the DB.
+// Hesitant to do DeleteAsync because that changes the
+// expected ordering
+func (bb *badgerDBBatch) Delete(key []byte) {
+	bb.db.Delete(key)
```
Comment (melekes) [inline db/badger_db.go]: no error handling?

File: db/badger_db.go [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+}
+
+// Write commits all batch sets to the DB
+func (bb *badgerDBBatch) Write() {
+	bb.entriesMu.Lock()
+	entries := bb.entries
+	bb.entries = nil
```
Comment (melekes) [inline db/badger_db.go]: what if somebody calls `set` on this batch after this? code will panic

File: db/badger_db.go (hunk 1 above)
Comment (odeke-em) [reply db/badger_db.go]: Thanks, I'll update that.

File: db/badger_db.go (hunk 2 above)
Comment (odeke-em) [reply db/badger_db.go]: I don't think so, db.NewDB uses NewBadgerDB and expects the directory making to be handled magically. Moving it out of NewBadgerDB would mean that we'd have to document separately that BadgerDB needs to have its directory made before-hand, and that would make it awkward.

File: db/badger_db.go (hunk 3 above)
Comment (odeke-em) [reply db/badger_db.go]: For some reason, I initially thought `badger.DefaultOptions` was a pointer but somewhere I noticed not and left that in, lol. I'll update it, thank you!

**Comment by ebuchman** [comment]
Cool! Can we base this on Alexis' branch (https://github.com/tendermint/tmlibs/pull/57) ? I'd love to see some benchmarks of performance vs our Go LevelDB

**Comment by melekes** [comment]
also, tests are failing with a missing dependency error.

**Comment by melekes** [comment]
> Cool! Can we base this on Alexis' branch (#57) ? I'd love to see some benchmarks of performance vs our Go LevelDB

alexis' branch was merged into develop. so this branch needs to be rebased against develop

**Comment by melekes** [comment]
ok. I've rebased the PR. 

@odeke-em could you address some of my comments above and TODO here https://github.com/tendermint/tmlibs/pull/58/commits/2a54ceb94ba88370257622bea337b19cd3d5f378. 

Also, I would love to see benchmarks too https://github.com/tendermint/tmlibs/pull/58#issuecomment-333722132

**Comment by melekes** [comment]
**I also have a concern that we're adding another dependency** (github.com/dgraph-io/badger) to our list of deps, which is already long. Is there a way we could only require it if it's being used?

**Comment by melekes** [comment]
```
db/badger_db.go:277:22:warning: should use buf.String() instead of string(buf.Bytes()) (S1030) (megacheck)
db/badger_db.go:282:2:warning: field mu is unused (U1000) (megacheck)
db/badger_db.go:282:2:warning: unused struct field github.com/tendermint/tmlibs/db.badgerDBIterator.mu (structcheck)
db/badger_db_test.go:88::error: wrong number of args for format in Errorf call: 2 needed but 3 args (vet)
db/badger_db_test.go:186::error: wrong number of args for format in Errorf call: 2 needed but 3 args (vet)
db/example_test.go:24::error: Example_BadgerDB has malformed example suffix: BadgerDB (vet)
```

**Comment by cloudhead** [comment]
If we're going to support a bunch of backends and don't want to pull in all the deps, we can do something like:

```
import "tmlibs/db/base"
import "tmlibs/db/drivers/badger"
import "tmlibs/db/drivers/leveldb"

func init() {
  base.Register("badger", badger.NewDriver())
  base.Register("leveldb", leveldb.NewDriver())
}
```

**Comment by odeke-em** [comment]
@melekes I mean do we want the DB though? If we do, I don't see how we can avoid adding the dependency.

@cloudhead what do you mean by "not pull  in all the deps"? By including all those imports, the respective object code has to be compiled and linked. Also I believe we already have that style of registering backends https://github.com/tendermint/tmlibs/blob/d9525c0fb671204450b160807480e1263053fb20/db/db.go#L50-L64 but I perhaps don't understand what the concern here is, please help me understand it.

**Comment by odeke-em** [comment]
@melekes thank you for the vets, I'll update them!

**Comment by odeke-em** [comment]
@melekes fist bump for the vets and suggestions to make badgerDBIterator return an error, I've updated those points in commit https://github.com/tendermint/tmlibs/pull/58/commits/534cd7b90667687a289d0558676bc9b6545b7425

For the benchmarks, this cafe is as cold as Siberia, let me walk home, grab dinner and work on them there.

File: db/badger_db.go (hunk 7 above)
Comment (odeke-em) [reply db/badger_db.go]: Nope, after this, it starts a new batch. Take a look at the code in Set https://github.com/tendermint/tmlibs/pull/58/files#diff-9a61ac3e18917559e9062e23d4e9435dR237
or
screen shot 2017-11-05 at 2 01 10 pm

File: db/badger_db.go (hunk 6 above)
Comment (odeke-em) [reply db/badger_db.go]: The 'error handling' present with our structure doesn't return an error yet I am apprehensive about adding panics in Delete.

**Comment by cloudhead** [comment]
@odeke-em sorry probably wasn't clear: the code snippet I wrote would happen user-side, not in tmlibs. So the user can import additional backends and register them with tmlibs/db. By default, tmlibs/db wouldn't come with any disk-based backends.

**Comment by greg-szabo** [comment]
On today's (2018 feb 5th) tendermint dev meeting Bucky mentioned that according to Bucky's and Jae's research, BadgerDB is not mature enough for us to be used. Shall this PR be closed? If not, what's the priority / use case?

**Comment by odeke-em** [comment]
Yeah @greg-szabo in deed, I raised some problems here https://github.com/tendermint/tmlibs/pull/58#issue-261880212 and also in the filed issue on BadgerDB https://github.com/dgraph-io/badger/issues/308, their API had drastically changed in the span of 2 weeks, but also that the max length of their keys is 1<<16 https://github.com/dgraph-io/badger/issues/308#issuecomment-342052570.

**Comment by greg-szabo** [comment]
Closing this PR. Thanks for the work on it odeke, even though we pass on BadgerDB, it was an important point to raise and discuss it.

**Pull request closed without merging**